### ISSUE TITLE
Visualize and edit graph like data in napari

### ISSUE THREAD

**Issue by ClementCaporal** [napari/napari#4274]
## 🚀 Feature
`graph-layer` to visualize and edit graph data in napari.

## Motivation

I have graph-like data such as neurons paths or vertex model of epithelium. I would like to visualize it with my favorite image viewer in Python. 
More generally I suppose that graph-like data such as twitter-network, molecules, ... could also use this layer type?

## Pitch

- create new graph-layer
- load existing graph from a graph library (such as [networkx](https://github.com/networkx/networkx)? or other)
- visualize in the viewer the graph and change the layout of edge/node
- edit the graph by deleting only one node or a sub-branch

## Alternatives

It is already possible to visualize graph data in napari using point, label or shape layer but is not optimal to edit.
- vector and label layer example: [skan](https://skeleton-analysis.org/visualizing_3d_skeletons.html)

`track-layer` also have utils/visualization that can be reused?

## Additional context
This issue follows a discussion with @alisterburt and @JoOkuma.

**Comment by tlambert03** [comment]
cc @jni ... who I know is quite interested / working on this as well

**Comment by brisvag** [comment]
Relevant: https://github.com/vispy/vispy/tree/main/vispy/visuals/graphs

**Comment by jo-mueller** [comment]
Would graphs in Napari  strictly be spatial graphs? Or in other words, would a node in a graph layer always have a location in nd-space?

**Comment by ClementCaporal** [comment]
> Would graphs in Napari strictly be spatial graphs? Or in other words, would a node in a graph layer always have a location in nd-space?

I think position should be an optional argument. It would bypass a default draw layout argument ([like for example in networkx](https://networkx.org/documentation/stable/reference/drawing.html#module-networkx.drawing.layout))

**Comment by jni** [comment]
Hi @ClementCaporal and @jo-mueller! Great that you are interested here! Indeed we have a number of people interested already. Here's notes from our first meeting:

https://hackmd.io/aDecpm9FR8qYAoal-7_1Lg

@JoOkuma is also interested I believe, because the vision is for Tracks to be a special-case of Graph.

Unfortunately none of us have had time to even start implementing this. If you have bandwidth to work on this I am very happy to help. Either way, I hope to reboot the Graph working group and will ping here when I set up the next meeting.

> > Would graphs in Napari strictly be spatial graphs? Or in other words, would a node in a graph layer always have a location in nd-space?
> 
> I think position should be an optional argument. It would bypass a default draw layout argument ([like for example in networkx](https://networkx.org/documentation/stable/reference/drawing.html#module-networkx.drawing.layout))

Honestly I find this a challenging question. I understand the convenience of not having to specify a layout, but napari needs to know coordinates in order to position the nodes *relative* to other layers. Without the layout, we don't even know if we want a 2D, 3D, or 4D layout! Or if the scale should be in [0, 1], [0, 512], or [0, 2**16]. All these things can of course be parameters to the layer, but at the moment it makes more sense to me to require users to call a layout algorithm *first*, then provide the output to the layer, rather than have the layer call it for us.

That's all a weak opinion, but it is my initial gut reaction. Happy to discuss further of course!

@ClementCaporal btw,

> I have graph-like data such as neurons paths

Have you seen my "manual" layout solution using skeleton graphs?

https://skeleton-analysis.org/visualizing_3d_skeletons.html

Still a lot of boilerplate so 100% we want to make this easier (also on my to-do...), but it might be useful for you in the meantime...

**Comment by ClementCaporal** [comment]
Thank you for the notes of you first meeting @jni .

> ## Alternatives
> 
> It is already possible to visualize graph data in napari using point, label or shape layer but is not optimal to edit.
> 
>     * vector and label layer example: [skan](https://skeleton-analysis.org/visualizing_3d_skeletons.html)


I did see your solution as specified in the alternative. I didn't test it yet but I will.
For now I use [Genepy3D](https://genepy3d.gitlab.io/) + Point Layer to compare my automatic tracing

- ![image](https://user-images.githubusercontent.com/94049435/158801551-d295e0ab-0dcf-42b7-9d62-ae9c2a98b6f3.png)
(Each layer is one automatic generated trace, the bottom row is the manual trace and the raw image)
- ![image](https://user-images.githubusercontent.com/94049435/158802157-6ec72457-870c-49b4-a08e-c71bd53ac275.png)
(Or I just superpose the layers)


I want to dedicate some time to help on this graph-layer too. I suppose I will wait for the next meeting.

Thank you!

**Comment by jo-mueller** [comment]
I would also be interested in contributing. From the perspective of my use cases, I would be fine with having more abstract graph representations (nodes and edges) to be able to freely pass criteria like neighborhood between objects (can be points or labels or whatever) between plugins, similar to what has been asked [here](https://forum.image.sc/t/interactive-visualisation-of-graphs/60629)

> > > Would graphs in Napari strictly be spatial graphs? Or in other words, would a node in a graph layer always have a location in nd-space?
> > 
> > 
> > I think position should be an optional argument. It would bypass a default draw layout argument ([like for example in networkx](https://networkx.org/documentation/stable/reference/drawing.html#module-networkx.drawing.layout))
> 
> Honestly I find this a challenging question. I understand the convenience of not having to specify a layout, but napari needs to know coordinates in order to position the nodes _relative_ to other layers.

I think I would stay away from coming up with graph layouts within the scope of a graphs layer. The [networkx graph drawing documentation ](https://networkx.org/documentation/stable/reference/drawing.html) says that "Proper graph visualization is hard, and we highly recommend that people visualize their graphs with tools dedicated to that task." so...I'd rather not go down this road, or simply use other library's function if no node position is passed.


> Have you seen my "manual" layout solution using skeleton graphs?
> 
> https://skeleton-analysis.org/visualizing_3d_skeletons.html

That's cool! I think reusing some of the implemented layer functionalities for graphs would be a feasible way to go to prevent re-inventing the wheel. I could imagine that it is feasible to implement a graph layer as a mixture of a points (nodes) and a vectors (edges) layer. 
One could then pass additional properties to each, such as an additional list of points that denote non-straight traces as edges.

**Comment by sofroniewn** [comment]
> at the moment it makes more sense to me to require users to call a layout algorithm first, then provide the output to the layer, rather than have the layer call it for us.

> I think I would stay away from coming up with graph layouts within the scope of a graphs layer. The [networkx graph drawing documentation ](https://networkx.org/documentation/stable/reference/drawing.html)says that "Proper graph visualization is hard, and we highly recommend that people visualize their graphs with tools dedicated to that task." so...I'd rather not go down this road, or simply use other library's function if no node position is passed.

I agree with both of the sentiments above and think it's more in line with the core of napari focusing on visualization and interactivity - we could imagine a plugin computing layouts for example, but the napari layer should just accept one.

Overall it looks like there's a lot of great ideas in this thread and interest and I'm wondering if there is anything I can do to help it along - will we need to add a dependency for a graph library here? or settle on the API of a graph library to match? that seems like a big choice ahead

**Comment by JoOkuma** [comment]
I think adding a graph dependency would help to implement the graph and tracks layer algorithms just like we call numpy functions inside the layers.

Some graph algorithms can't be implemented using matrix operations and are very slow in pure python, by having a dependency that allows cython (or something similar) we can contribute to them and avoid adding this to napari.

I also support the graph requiring spatial coordinates. However, it's not clear if that information would be in the graph structure or using the layer features.

**Comment by jo-mueller** [comment]
> I think adding a graph dependency would help to implement the graph and tracks layer algorithms just like we call numpy functions inside the layers.

Isn't networkx already part of the Napari dependencies?

**Comment by brisvag** [comment]
> I also support the graph requiring spatial coordinates. However, it's not clear if that information would be in the graph structure or using the layer features.

I think that features should be for optional/additional data, not something required, so I'd say it should definitely be a first class citizen. I don't really see a way to handle graphs without spatial coordinates in napari; what happens if you display it together with other data?

Sure we can generate a structure ourselves, but as pointed out by @jo-mueller, this is not trivial and should be probably handled by plugins or other libraries.

It should be relatively easy to use networkx attributes for positions. Something that might be trickier is how to handle Features in a way that works "natively" with graphs algorithms, while exposing them as a `FeatureTable` on the napari side (cc. @andy-sweet).

**Comment by JoOkuma** [comment]
> @jo-mueller I noticed that too when updating napari.

I noticed that too when updating napari.

> I think that features should be for optional/additional data, not something required, so I'd say it should definitely be a first class citizen.
> It should be relatively easy to use networkx attributes for positions. Something that might be trickier is how to handle Features in a way that works "natively" with graphs algorithms, while exposing them as a FeatureTable on the napari side (cc. @andy-sweet).

If we opt for NetworkX every feature could be saved on the nodes (what happens to the edges?) attributes. Saving in a table makes things awkward when adding and removing nodes, this is a current issue with the tracks layers.

My major concern with NetworkX is that it's slower than pure python.

**Comment by brisvag** [comment]
> My major concern with NetworkX is that it's slower than pure python.

Oh, I didn't know about that! I'm definitely not equipped to have an opinion about which library to use :P

> If we opt for NetworkX every feature could be saved on the nodes (what happens to the edges?) attributes. Saving in a table makes things awkward when adding and removing nodes, this is a current issue with the tracks layers.

Yeah, this is what I though as well, but it would also be best to have the `Feature` experience match between layer types. How to handle edges is another beast... Maybe having two separate `Feature`s is the most straightforward?

**Comment by jo-mueller** [comment]
> If we opt for NetworkX every feature could be saved on the nodes (what happens to the edges?) attributes.

I agree with this approach. Actually, I would re-use the points layer for this to a large degree. The nodes in the graph would have to have an attribute `coordinates` which could then directly be converted to a points layer under the hood.

The edges in networkx more or less just connect, for instance, `point 1` to `point 2` so that the edge could simply be visualized with a napari vector, i.e. `point_1['coordinates'] - point_2['coordinates']`. Edges could also have an optional attribute `coordinates` (similar to `coordinates` in nodes) which could indicate non-straight connections between nodes.

**Comment by brisvag** [comment]
I agree that the visualization should use vispy's `Markers` (same as points) and the lines used by vectors. However, keep in mind that on the `ViewerModel` side (that is, where `Layers` themselves live), we only work with the data, not with the visualization itself, and I think we should focus on staying as true to the data as possible (a `Graph`) rather than using a mix of other layers.

Otherwise, we can simply use two separate layers and a layergroup, once those are merged... but it seems to me that a `Graph` layer would require something more "native" than that, in order to take advantage of all the graph theory stuff :P

**Comment by JoOkuma** [comment]
I can implement non-interactive graph layer so I can move forward with the tracks layer rework with a similar API. 

What are the appropriate next steps for that? Should I open a discussion to decide on the data format / graph library?

@jni @sofroniewn

**Comment by sofroniewn** [comment]
> What are the appropriate next steps for that? Should I open a discussion to decide on the data format / graph library?

yeah - that might be helpful. I'm not sure if this is the sort of thing that would need to go through a NAP (see #4299) but that process might help here.

I think it would be good to have a list of the different possible options, along with the pros/ cons of each and then if you have one your recommendation on which to pursue

**Comment by JoOkuma** [comment]
I started working on a document for us to decide on the graph data type, please leave your comments.

https://hackmd.io/tiltu-J8QBOj8TwnlkMgAQ

cc @jni @jwindhager @quantumjot

**Comment by jwindhager** [comment]
Thanks for tagging me @JoOkuma 

I planned to start working on a graph layer end of June - won't be sad if others have implemented it by then :-)

After our initial meeting with @jni @DragaDoncila I had the following corner stones in mind:
- **Graph storage**: for efficiency (i.e., not networkx) and staying pythonic (i.e., no igraph), I was thinking about storing the graphs in numpy arrays/pandas DataFrames, without additional dependencies. These representations can be easily converted to networkx/igraph objects, potentially even without data duplication, ensuring interoperability with popular graph libraries. Thinking about it again, I realize that this might not be the optimal strategy though, as adding/removing single nodes would become very expensive operations (numpy array shape changes) in this strategy. But I would still vote for a package-agnostic way of storing graphs, so maybe it's still helpful to document my thoughts somewhere:
  - Adjacency matrix: boolean numpy array or mutable scipy sparse matrix, potentially with memory optimization for storing undirected graphs (i.e., symmetric adjacency matrices); caveat: numpy stores boolean as 8 bit
  - Node attributes: pandas DataFrame, with one column per attribute; index contains node IDs
  - Edge attributes: multiple scipy sparse matrices (one matrix per attribute); alternatively, make the adjacency matrix non-boolean, storing edge IDs (!= 0), which can then become the index in an edge attribute pandas DataFrame
  - Spatial coordinates can (optionally!) be stored as regular node attributes (see graph layouts below)
  - Given the relationships between adjacency matrices, node attribute tables and edge attribute matrices, one idea would also be to leverage xarray's `Dataset` storing multiple `DataArray` instances with labels shared among them. However, afaik DataArray's currently do not support the same coordinate on both axes of a square matrix (e.g. adjacency matrices), and xarray is currently not a dependency of napari
  - Multigraphs: multiple adjacency matrices?
  - Hypergraphs, bipartite graphs, ...?
- **Graph file formats**: provide reader/writer functionality for formats supported by networkx/igraph (if installed)
- **Graph visualization**: 
  - The vispy Graph visual was already mentioned above
  - Graph layouts: I would support the visualization of both Euclidean and non-Euclidean graphs. Similar to e.g. CytoScape or gephi, I would let the user choose from a variety of graph layouts, including both automatic and "data-driven" graph layouts, where for the latter the spatial node coordinates are directly from user-defined node attributes (see above).

**Comment by jni** [comment]
I want to add an additional option for the graph data structure, which is the [trackmate graph](https://github.com/mastodon-sc/mastodon/blob/master/doc/trackmate-graph.pdf), implemented in java [here](https://github.com/mastodon-sc/mastodon-graph) (I think that's the right implementation link). @DragaDoncila thinks that it wouldn't be too hard to implement in Python + numba. To me it feels like a lot of effort but not an insurmountable amount, and of all the solutions so far it is the only one that has been battle tested in this exact use case. Does anyone have the bandwidth to work on this directly? We can make a napari/graph repo and start work there.

I also want to reiterate that both @sofroniewn and I think automatic generation of graph layouts is out of scope for napari. The Graph layer input *must* include node coordinates. Edge paths are optional but should also be supported.

**Comment by jni** [comment]
I've created https://github.com/napari/napari-graph where we can iterate quickly on a Graph data object (more quickly than we could do in this repo). Once we have an object that supports the operations we want at some reasonable speed, I think it would be reasonably straightforward both to make a Graph layer and to make Tracks use the Graph object internally also.

**Comment by jwindhager** [comment]
> I also want to reiterate that both @sofroniewn and I think automatic generation of graph layouts is out of scope for napari. The Graph layer input _must_ include node coordinates. Edge paths are optional but should also be supported.

I fully agree that napari shouldn't implement or even come up with new automatic graph layouts. But since we anyway need a user interface for mapping node attributes to spatial positions (if only to support direct loading graph file formats), why not ship napari with a single "data-driven" layout (thin "Proxy" class mapping node coordinate attributes to napari dimensions) and let the addition of further layouts up to plugins?

**Comment by jni** [comment]
> But since we anyway need a user interface for mapping node attributes to spatial positions

I disagree that we need such a thing. napari is an *image* viewer, and everything else is in support of that primary goal. We have a points layer not because we want to be a plotting library, but because it is useful to annotate images with points. We are not and will never be (as far as I'm concerned) a generic graph display library. Loading a graph file that does not already contain spatial positions is therefore meaningless in napari.

If someone wants to repurpose our infrastructure to write a plugin that enables the display of abstract graphs (without intrinsic spatial information), that's super great, that is one of the awesome powers of open source, but it should be maintained outside of our organisation.

> why not ship napari with a single "data-driven" layout

Because such things, however small, involve a maintenance cost, and the maintenance cost of napari is already huge.

**Comment by jwindhager** [comment]
Alright, fair point, no graph layouts then. Just noting here that this means that potential graph reader/writer plugins will have to deal with this issue, either through a GUI (note: separate GUIs for each plugin) or by imposing additional attribute naming constraints on the graph files (as there doesn't seem to be a standard naming scheme for spatial coordinate attributes).

**Comment by JoOkuma** [comment]
I have the bandwidth to work on this intermittently.

We could start with the graph data structure and the interface with network x, and implement the algorithms as we go.

I'm in favor of graph with nodes indexed with a integer dictionary and edges stored as lists for each node.

We can list this and other data structure options on the hackmd document. And this doesn't need to be set in stone, we can change it later as long as our API is consistent.

**Comment by JoOkuma** [comment]
I agree with @jwindhager that we would need some naming conventions for spatial coordinates on features, I much rather have this than the column order with Tracks and we're already using pandas anyway.

**Comment by jni** [comment]
@JoOkuma 

> We could start with the graph data structure and the interface with network x,

*which* graph data structure?

But yes, I agree with creating an API shell that mimics networkx, regardless of our internal data structure. Indeed, perhaps we can do several "views" of our internal data, NGraph.nx, NGraph.csr_array, etc.

**Comment by JoOkuma** [comment]
@jni 

> which graph data structure?

I'm not sure which one is best yet. We could choose based on the "Desirable Properties" section of the document, I added a new item to it.

I would start with the goal of having a graph layer that supports an undirected graph with no parallel edges. Later, we could add the option to scale the edge weight given a feature. I don't what's the expected behavior for the nodes, should they be scalable by feature too?

For tracking we'll have a directed graph with a few special properties, it needs to be sliced over time efficiently, limited number of dimensions (?), and without cycles. It should inherit the undirected graph class.

**Comment by jni** [comment]
@JoOkuma it's unclear to me above whether you've read my comments about the trackmate graph, and the trackmate graph "white paper" itself?

**Comment by jwindhager** [comment]
Just want to say that IMHO the trackmate graph structure (linked edge lists, essentially) looks very suitable for our purposes :+1: I propose to store node/edge attributes outside of that data structure, though, and link nodes/edges to node attributes/edge attributes with an ID - this way we decouple the tabular data structures from the graph topology, which makes it easier to change them in the future.

Regarding *which* data structure to use for node/edge attributes: pandas DataFrames come to mind, with obvious drawbacks wrt. adding/removing nodes/edges. However, for consistency with `features` in other layer types, we could start with that (I'm not up-to-date whether there are any plans to replace DataFrames in the near future?).

**Comment by jwindhager** [comment]
> I would start with the goal of having a graph layer that supports an undirected graph with no parallel edges.

Just my 2C: considering that we focus on Euclidean graphs, which are often some sort of kNN-graphs, I think that support for directed graphs would be important. I'd agree with parallel edges (i.e., multigraphs) not being too important at this point, but with the trackmate graph structure proposed by @jni they come for free :-)

**Comment by JoOkuma** [comment]
@jni Sorry, I did miss it.

I agree with storing vertices in an array and marking removed elements, this could speed things up when iterating every node of the graph in any order. We could use a node data class and not some fixed memory space so we could store an arbitrary number of features. Or it could even be the case to store each feature as its own array and use the same index and delete policy as the vertex array.

However, it isn't clear the benefits of having the edges as an array, you have no guarantee that edges are near each other when iterating, so you can't benefit from data locality. I guess that implementing a simple linked list (without storing an array) should yield similar results, as @jwindhager suggested.

I have the impression that their data structure is optimized to avoid data allocation. I'm not too worried about that, I think having incremental addition/removal of edges in constant time is important (this is not the case for the current tracks layer) but we don't need a complex implementation, because this would happen the most when the user is interacting and allocating memory is not that expensive to impair the user interaction.

Being very fast to iterate over the graph is the most important goal in my opinion because that's what we need to do to display it faster in napari if the graph needs to be sliced. This could be achieved by having a data structure that's fast to index, like the array of vertices.

**Comment by JoOkuma** [comment]
As I think more about it, storing vertices' coordinates and features in arrays and marking deleted regions could be very interesting, we could implement a shader to ignore these nodes, so we don't need to preprocess it to send it to vispy.

Maybe there's something similar that could be done with edges. I need to take a look at how the vispy graph visualizations are implemented.

**Comment by jni** [comment]
> However, it isn't clear the benefits of having the edges as an array, you have no guarantee that edges are near each other when iterating, so you can't benefit from data locality. I guess that implementing a simple linked list (without storing an array) should yield similar results, as @jwindhager suggested.

This is true in general, but not true in two very important graphs: tracks and skeletons. In both of these, tracks and paths usually only have one neighbour and depending on how the graph is built, they could very much be immediately adjacent. They would therefore yield significant performance improvements when iterating compared to a linked list. At any rate, I suspect that is the first thing the trackmate folks tried, so we should assume they have good reason for using the edge list structure they use.

The other important thing is that data allocation is "hard" in Numba/Cython etc. It might ultimately be better to allocate only rarely in pure Python and then work with the pre-allocated structures. I'd prefer to avoid dealing with C-level memory management in napari altogether. :sweat_smile:

**Comment by jwindhager** [comment]
> However, it isn't clear the benefits of having the edges as an array, you have no guarantee that edges are near each other when iterating, so you can't benefit from data locality. I guess that implementing a simple linked list (without storing an array) should yield similar results, as @jwindhager suggested.

(just for clarity, I didn't suggest to NOT use arrays in the implementation)

**Comment by JoOkuma** [comment]
@jwindhager I thought you meant a simple linked list.

> At any rate, I suspect that is the first thing the trackmate folks tried, so we should assume they have good reason for using the edge list structure they use.

I have a very experimental implementation at https://github.com/napari/napari-graph/pull/1.
The trackmate-like structure (`UndirectedGraphBufferLLArray`) is the fastest so far. Even though there are a few operations to be implemented I think it might be the fastest data structure, despite requiring resizing the arrays once in a while.

I had forgotten how slow python is, numba is doing the trick :smile: .

**Comment by jni** [comment]
> I had forgotten how slow python is, numba is doing the trick

:rocket:

**Comment by JoOkuma** [comment]
Hi, everyone.

I implemented an initial version of the [napari-graph](https://github.com/napari/napari-graph). With this implementation, we'll speed up the graph initialization and it'll allow us to manipulate the graph.

Beyond that, I think it might be doable to have a single layer for graphs and tracks, where tracks are a directed graph with a sliding window (thick slices) over the time dimension.

I'm going to start working soon on the graph layer and do the final changes to the initial graph data structures of `napari-graph`.

Anyone interested in napari's graph layer is welcome to review [napari-graph#1](https://github.com/napari/napari-graph/pull/1).

**Comment by alisterburt** [comment]
Really exciting @JoOkuma ! Would love to see a demo soon :)

**Comment by brisvag** [comment]
> Beyond that, I think it might be doable to have a single layer for graphs and tracks, where tracks are a directed graph with a sliding window (thick slices) over the time dimension.

<3 I think this shouldn't be a special case for tracks; we should be able to make it part of the general `thick slices` implementation.

**Comment by sofroniewn** [comment]
Awesome @JoOkuma - great to hear about progress here!!

@quantumjot you might find this exciting! if you or any of your team want to check napari-graph out, see this comment https://github.com/napari/napari/issues/4274#issuecomment-1166144787
